framework/keyring-interface.md

```diff
@@ -36,37 +36,48 @@ in this document are to be interpreted as described in [RFC 2119](https://tools.
 
 ### OnEncrypt
 
-This interface takes [encryption materials](#structures.md#encryption-materials) as input
-and MAY modify it with any of the following behaviors:
+The following inputs are REQUIRED:
+
+- [Algorithm Suite](#algorithm-suites.md)
+- [Encryption Context](#encryption-context)
+
+The following inputs are OPTIONAL:
+
+- [Plaintext Data Key](#structures.md#plaintext-data-key)
+
+This interface MAY return [data key materials](#structures.md#data-key-materials) appropriate for the request. If it does return this output
+and a plaintext data key was provided as input, the output must specify the same plaintext data key.
+
+The output of this interface is driven by two behaviours:
 
 - [Generate data key](#generate-data-key)
 - [Encrypt data key](#encrypt-data-key)
 
 If this keyring attempted any of the above behaviors, and successfully completed those behaviors,
-it MUST output the modified [encryption materials](#structures.md#encryption-materials).
+it MUST output the result as [data key materials](#structures.md#data-key-materials).
 
-If the keyring did not attempt any of the above behaviors, it MUST output the
-[encryption materials](#structures.md#encryption-materials) unmodified.
+If the keyring did not attempt any of the above behaviors, it MUST produce no output.
```

Comment:
"The interface MUST fail" is the equivalent language for throwing an exception. I'm using "produce no output" to mean produce null or None or something equivalent rather than a result or Some(something). We need something like this to describe interfaces that are allowed to "do nothing" A keyring can decide to not attempt to decrypt any encrypted data keys, but that's different than attempting and failing.

Reply:
What if it threw something like a EncryptionNotRequiredException? Just trying to see if we can avoid returning null. It could return Optional<DataKeyMaterials> but I'm not sure I love that either

Reply:
Nah, not a fan of that since it's encouraging exceptions-as-control-flow. This isn't an exceptional situation so it shouldn't produce an exception. :)

Note that the abstract descriptions of behaviour can map to actual implementation in non-trivial ways. "produce no output" might be "don't populate any of the out parameters" if that is unambiguous (and that's what the older version was dictating in a more restrictive way).

Reply:
> Nah, not a fan of that since it's encouraging exceptions-as-control-flow. This isn't an exceptional situation so it shouldn't produce an exception. :)

This is a language-specific design choice and as such should not be enshrined in the specification. This is why we used the "must fail" terminology; it leaves the decision to the implementor how specifically to manifest "failure" in their implementation.

> Note that the abstract descriptions of behaviour can map to actual implementation in non-trivial ways. "produce no output" might be "don't populate any of the out parameters" if that is unambiguous (and that's what the older version was dictating in a more restrictive way).

We should not be intentionally opening the spec to interpretation in this way. "How do I fail" is a very different thing than "maybe I'll emit something, maybe I won't, maybe that means something, maybe it doesn't".

Reply:
I'm trying to introduce "produce no output" as distinct behaviour from "fail". In functional style this would mean, for example, returning an `Option<T>`. "producing no output" would map to returning `None`, but "failing" would make to raising an exception. Or in some cases you would return an `Either<Option<T>, Error>`, in which case "producing no output" would mean returning `Left(None)` whereas "failing" would mean returning `Right(Error(...))`.

I completely agree that we shouldn't dictate how either behaviour is implemented, but they must be implemented as distinguishable behaviours somehow.